drivers/sdhc/sdhc_renesas_ra.c

```diff
@@ -0,0 +1,743 @@
+extern fsp_err_t r_sdhi_hw_cfg(sdhi_instance_ctrl_t *const p_ctrl);
+extern fsp_err_t r_sdhi_read_and_block(sdhi_instance_ctrl_t *const p_ctrl, uint32_t command,
+				       uint32_t argument, uint32_t byte_count);
+extern fsp_err_t r_sdhi_wait_for_device(sdhi_instance_ctrl_t *const p_ctrl);
+extern fsp_err_t r_sdhi_wait_for_event(sdhi_instance_ctrl_t *const p_ctrl, uint32_t bit,
+				       uint32_t timeout);
+extern void r_sdhi_command_send_no_wait(sdhi_instance_ctrl_t *p_ctrl, uint32_t command,
+					uint32_t argument);
+extern void r_sdhi_read_write_common(sdhi_instance_ctrl_t *const p_ctrl, uint32_t sector_count,
+				     uint32_t sector_size, uint32_t command, uint32_t argument);
+
+struct sdhc_ra_config {
+	const struct pinctrl_dev_config *pcfg;
+	void *const regs;
+};
+
+struct sdhc_ra_priv {
+	struct st_sdmmc_instance_ctrl sdmmc_ctrl;
+	struct st_sdmmc_cfg fsp_config;
+	struct gpio_dt_spec sdhi_en;
+	struct sdmmc_ra_event sdmmc_event;
+	uint8_t channel;
+	bool app_cmd;
+	uint32_t bus_clock;
+	uint8_t bus_width;
+	enum sdhc_timing_mode timing;
+	enum sdhc_power power_mode;
+	struct k_sem thread_lock;
+	uint8_t status;
+	struct sdhc_host_props props;
+	/* Transfer DTC */
+	struct st_transfer_instance transfer;
+	struct st_dtc_instance_ctrl transfer_ctrl;
+	struct st_transfer_info transfer_info;
+	struct st_transfer_cfg transfer_cfg;
+	struct st_dtc_extended_cfg transfer_cfg_extend;
+};
+
+void sdhimmc_accs_isr(void);
+void sdhimmc_card_isr(void);
+void sdhimmc_dma_req_isr(void);
+
+static void ra_sdmmc_accs_isr(const void *parameter)
+{
+	ARG_UNUSED(parameter);
+	sdhimmc_accs_isr();
+}
+
+static void ra_sdmmc_card_isr(const void *parameter)
+{
+	ARG_UNUSED(parameter);
+	sdhimmc_card_isr();
+}
+
+static void ra_sdmmc_dma_req_isr(const void *parameter)
+{
+	ARG_UNUSED(parameter);
+	sdhimmc_dma_req_isr();
+}
+
+static int sdhc_ra_get_card_present(const struct device *dev)
+{
+	struct sdhc_ra_priv *priv = dev->data;
+	fsp_err_t fsp_err;
+	int ret;
+	sdmmc_status_t status;
+
+	/* SDMMC_CARD_DETECT_CD must be configured as true to check here */
+	fsp_err = R_SDHI_StatusGet(&priv->sdmmc_ctrl, &status);
+	ret = err_fsp2zep(fsp_err);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return (status.card_inserted);
+}
+
+static int sdhc_ra_card_busy(const struct device *dev)
+{
+	struct sdhc_ra_priv *priv = dev->data;
+	fsp_err_t fsp_err;
+	int ret;
+	sdmmc_status_t status;
+
+	fsp_err = R_SDHI_StatusGet(&priv->sdmmc_ctrl, &status);
+	ret = err_fsp2zep(fsp_err);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return (status.transfer_in_progress);
+}
+
+static int sdhi_command_send_wait(sdhi_instance_ctrl_t *p_ctrl, uint32_t command, uint32_t argument,
+				  uint32_t timeout)
+{
+	/* Verify the device is not busy. */
+	r_sdhi_wait_for_device(p_ctrl);
+
+	/* Send the command. */
+	r_sdhi_command_send_no_wait(p_ctrl, command, argument);
+
+	/* Wait for end of response, error or timeout */
+	return r_sdhi_wait_for_event(p_ctrl, SDHI_PRV_RESPONSE_BIT, timeout);
```

Comment:
I'd still like to see this addressed- this implementation is polling on the status of the response bit: https://github.com/zephyrproject-rtos/hal_renesas/blob/31f1d1b6c78b6d1e2dbf7a63fdc71e7ba6a74740/drivers/ra/fsp/src/r_sdhi/r_sdhi.c#L2137

Does the hardware provide an interrupt that could be used here instead? If you'd prefer to keep this polling you may (it is your driver after all), but I'd prefer to use interrupts here if we can.

Reply:
Our hardware doesn't provide a specific interrupt that could be used to check the response status bit. The response bit will combine all events for each command. May I keep the polling here?

Reply:
Of course- again, it is your driver, this was just a suggestion!